File: app/src/main/java/euphoria/psycho/library/BookListActivity.java
package euphoria.psycho.library;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.*;
import android.widget.*;

import java.util.List;


public class BookListActivity extends Activity implements AdapterView.OnItemClickListener {
    private static final int MENU_ADD_CLIPBOARD = 0;
    private static final int MENU_CHANGE_TAG = 1;
    private static final String TAG = "BookListActivity";
    private BookAdapter mBookAdapter;
    private List<String> mList;
    private ListView mListView;

    private void addFromClipboard(String s) {
        ClipboardManager clipboardManager = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);

        ClipData clipData = clipboardManager.getPrimaryClip();
        if (clipData != null && clipData.getItemCount() > 0) {
            CharSequence prefix = clipData.getItemAt(0).getText();
            if (prefix != null) {
                DataProvider.getInstance().addFromClipboard(s, prefix.toString());
                Toast.makeText(this, String.format("成功从剪切板添加%s.", s), Toast.LENGTH_SHORT).show();
                // refreshList();
            }
        }
    }

    private void addFromClipboard() {
        ClipboardManager clipboardManager = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);

        ClipData clipData = clipboardManager.getPrimaryClip();
        if (clipData != null && clipData.getItemCount() > 0) {
            CharSequence prefix = clipData.getItemAt(0).getText();
            if (prefix != null) {
                DataProvider.getInstance().insertArticle(prefix.toString());
                refreshList();
                Toast.makeText(this,"成功从剪切板添加.", Toast.LENGTH_SHORT).show();

            }
        }
    }

    private void changeTag(final String s) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        final EditText editText = new EditText(this);
        editText.setText(s);
        builder.setView(editText);
        builder.setNegativeButton("取消", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                dialogInterface.dismiss();
            }
        });

        builder.setPositiveButton("确定", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                if (editText.getText() == null) return;
                DataProvider.getInstance().updateTag(s, editText.getText().toString());
                refreshList();
            }
        });
        builder.show();
    }

    private void refreshList() {
        mBookAdapter.switchData(DataProvider.getInstance().listTag());
    }

    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
        Intent intent = new Intent();
        intent.putExtra(MainActivity.KEY_TAG, mList.get(i));
        setResult(RESULT_OK, intent);
        finish();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.booklis_activity);
        mList = DataProvider.getInstance().listTag();
        mBookAdapter = new BookAdapter();
        mListView = findViewById(R.id.listView);
        mListView.setAdapter(mBookAdapter);
        mListView.setOnItemClickListener(this);
        registerForContextMenu(mListView);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        menu.add(0, MENU_ADD_CLIPBOARD, 0, R.string.add_from_clipboard);

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case MENU_ADD_CLIPBOARD:
                addFromClipboard();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
        menu.add(0, MENU_ADD_CLIPBOARD, 0, R.string.add_from_clipboard);
        menu.add(0, MENU_CHANGE_TAG, 0, R.string.change_tag);
        super.onCreateContextMenu(menu, v, menuInfo);
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        AdapterView.AdapterContextMenuInfo menuInfo = (AdapterView.AdapterContextMenuInfo) item.getMenuInfo();

        switch (item.getItemId()) {
            case MENU_CHANGE_TAG:
                changeTag(mList.get(menuInfo.position));
                return true;
            case MENU_ADD_CLIPBOARD:
                addFromClipboard(mList.get(menuInfo.position));
                return true;
        }
        return super.onContextItemSelected(item);
    }

    private class ViewHolder {
        public TextView title;
    }

    private class BookAdapter extends BaseAdapter {


        @Override
        public int getCount() {
            return mList.size();
        }

        @Override
        public String getItem(int i) {
            return mList.get(i);
        }

        @Override
        public long getItemId(int i) {
            return i;
        }

        @Override
        public View getView(int i, View view, ViewGroup viewGroup) {
            ViewHolder viewHolder;
            if (view == null) {
                view = LayoutInflater.from(BookListActivity.this).inflate(R.layout.list_item, viewGroup, false);
                viewHolder = new ViewHolder();
                viewHolder.title = view.findViewById(R.id.title);
                view.setTag(viewHolder);
            } else {

                viewHolder = (ViewHolder) view.getTag();
            }
            viewHolder.title.setText(mList.get(i));
            return view;
        }

        public void switchData(List<String> ls) {
            mList.clear();
            mList.addAll(ls);
            notifyDataSetChanged();
        }
    }

}

File: app/src/main/java/euphoria/psycho/library/BookListActivity.kt
package euphoria.psycho.library

import android.app.Activity
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Environment
import android.os.Handler
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.ContextMenu
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_book.*
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import okhttp3.OkHttpClient
import okhttp3.Request
import org.jetbrains.anko.coroutines.experimental.bg
import org.jetbrains.anko.db.NULL
import org.jetbrains.anko.find
import org.jsoup.Jsoup
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.TimeUnit

class BookListActivity : AppCompatActivity(), ToolbarManager {
    override val toolbar by lazy { find<Toolbar>(R.id.toolbar) }
    private val mHandler = Handler();
    var mBookListAdapter: BookListAdapter? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val unCaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler(object : Thread.UncaughtExceptionHandler {
            override fun uncaughtException(p0: Thread?, p1: Throwable?) {
                p1?.logToFile()
                if (unCaughtExceptionHandler != null) {
                    unCaughtExceptionHandler.uncaughtException(p0, p1)
                } else System.exit(2)
            }

        })
        setContentView(R.layout.activity_book)
        toolbar.inflateMenu(R.menu.menu_main)
        toolbar.setOnMenuItemClickListener {
            when (it.itemId) {
                R.id.action_add_from_clipboard -> {
                    addFromClipboard()
                    refreshListView()
                }
                R.id.action_add_from_network -> {
                    dialog("", "输入相应的网址") {
                        addFromURLImplement(it, "Network")
                        refreshListView()

                    }
                }
                else -> App.instance.toast("Unknown option")
            }
            true
        }

        bookList.layoutManager = LinearLayoutManager(this)
        attachTOScroll(bookList)
        enableHomeAsUp { finish() }
        registerForContextMenu(bookList)
    }

    override fun onCreateContextMenu(menu: ContextMenu?, v: View?, menuInfo: ContextMenu.ContextMenuInfo?) {
        menu?.apply {
            add(0, MENU_ADD_CLIPBOARD, 0, R.string.add_from_clipboard)
            add(0, MENU_ADD_FROM_URL, 0, "从网络添加")
            add(0, MENU_REPLACE_FROM_CLIPBOARD, 0, "从剪切板替换")
            add(0, MENU_CHANGE_TAG, 0, R.string.change_tag)
            add(0, MENU_DELETE_TAG, 0, R.string.context_menu_delete)
            add(0, MENU_EXPORT_FILE, 0, "导出")
        }

        super.onCreateContextMenu(menu, v, menuInfo)
    }

    fun addFromClipboard(tag: String) {
        val clipboardManager = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        if (!clipboardManager.hasPrimaryClip()) return
        if (clipboardManager.primaryClip.itemCount > 0) {
            val text = clipboardManager.primaryClip.getItemAt(0).text;
            if (text != null) {
                DataProvider.getInstance().addFromClipboard(tag, text.toString())
                val message = "成功添加：%s".format(tag);
                toast(message)
            }
        }
    }

    fun replaceFromClipboard(tag: String) {
        dialog("", "输入相应位置") {
            it.toIntOrNull()?.let {
                val clipboardManager = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                if (!clipboardManager.hasPrimaryClip()) return@let
                if (clipboardManager.primaryClip.itemCount > 0) {
                    val text = clipboardManager.primaryClip.getItemAt(0).text;
                    if (text != null) {
                        DataProvider.getInstance().updateDocument(tag, it, text.toString())
                        val message = "成功添加：%s".format(tag);
                        toast(message)
                    }
                }
            }
        }
    }

    fun changeTag(tag: String) {
        dialog(tag, "修改书本名") {
            if (it.isNotBlank())
                DataProvider.getInstance().updateTag(tag, it.trim())
            refreshListView();

        }

    }

    fun deleteByTag(tag: String) {
        DataProvider.getInstance().deleteByTag(tag);
        refreshListView();
    }

    private fun refreshListView() {
        mBookListAdapter?.switchData(DataProvider.getInstance().listTag())
    }

    override fun onContextItemSelected(item: MenuItem?): Boolean {

        var position = -1
        try {
            position = mBookListAdapter!!.position


        } catch (ex: Exception) {
        }
        when (item?.itemId) {
            MENU_ADD_CLIPBOARD -> {
                if (position != -1)
                    addFromClipboard(mBookListAdapter!!.getBook(position))
                return true
            }
            MENU_CHANGE_TAG -> {
                if (position != -1)
                    changeTag(mBookListAdapter!!.getBook(position))
                return true
            }
            MENU_REPLACE_FROM_CLIPBOARD -> {
                if (position != -1)
                    replaceFromClipboard(mBookListAdapter!!.getBook(position))
                return true
            }
            MENU_DELETE_TAG -> {
                if (position != -1)
                    deleteByTag(mBookListAdapter!!.getBook(position))
                return true
            }
            MENU_ADD_FROM_URL -> {
                addFromURL(mBookListAdapter!!.getBook(position))
                return true
            }
            MENU_EXPORT_FILE -> {
                mBookListAdapter?.let {
                    val book = it.getBook(position);
                    val dataProvider = DataProvider.getInstance();
                    val count = dataProvider.queryCount(book);
                    val parent = File(Environment.getExternalStorageDirectory(), ".readings");
                    val dir = File(parent, "exports");
                    dir.mkdirs()
                    val out = FileOutputStream(File(dir, "$book.txt"))

                    for (i in 0..count) {
                        
                        out.write(dataProvider.queryContent(book, i).toByteArray())
                    }
                    out.close()
                }
                return true
            }
            else -> return true
        }

    }

    fun addFromURL(tag: String) {

        dialog("", "输入相应的网址") {
            addFromURLImplement(it, tag)
        }

    }


    fun addFromURLImplement(url: String, tag: String) {
        if (url.contains("literotica")) {
            addFromLitErotica(url, tag)
        } else {
            async(UI) {
                var result = bg {

                    url.fetchString()

                }
                updateDatabase(tag, result.await())
                toast("成功添加：$url")
            }
        }

    }


    fun addFromLitErotica(url: String, tag: String) {
        Thread(Runnable {
            url.fetchHtml()?.let {

                val str = it.htm2txt(true);
                updateDatabase(tag, str);

                val js = Jsoup.parse(it)
                val elements = js.select(".b-pager-pages option")

                if (elements.size >= 0) {
                    val count = elements.size;
                    for (index in 2..count) {
                        val sublink = "$url?page=$index"
                        val value = "$index\r\n${sublink.fetchHtml()?.htm2txt(true)}"
                        updateDatabase(tag, value);
                    }
                }
                mHandler.post {
                    Toast.makeText(this@BookListActivity, url, Toast.LENGTH_LONG).show();
                }

            }


        }).start();
    }

    fun updateDatabase(tag: String, content: String?) {

        content?.let {
            DataProvider.getInstance().addFromClipboard(tag, content)
        }

    }

    fun addFromClipboard() {
        val clipboardManager = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        if (!clipboardManager.hasPrimaryClip()) return
        if (clipboardManager.primaryClip.itemCount > 0) {
            val text = clipboardManager.primaryClip.getItemAt(0).text;
            if (text != null) {
                DataProvider.getInstance().insertArticle(text.toString())
                toast("成功从剪切板添加.")
            }
        }
    }


    override fun onResume() {
        super.onResume()
        loadBook()
    }

    fun loadBook() {
        mBookListAdapter = BookListAdapter(DataProvider.getInstance().listTag()) {

            val intent = Intent().apply {
                putExtra(MainActivity.KEY_TAG, it)
            }
            setResult(Activity.RESULT_OK, intent)
            finish()
        }
        bookList.adapter = mBookListAdapter

    }

    companion object {
        private val MENU_ADD_FROM_URL = 3
        private val MENU_REPLACE_FROM_CLIPBOARD = 5
        private val MENU_ADD_CLIPBOARD = 0
        private val MENU_CHANGE_TAG = 1
        private val MENU_DELETE_TAG = 2

        private val MENU_EXPORT_FILE = 6
        private val TAG = "BookListActivity"
    }
}

